### ISSUE TITLE
needless-return for a match

### ISSUE THREAD

**Issue by yellowhat** [rust-lang/rust-clippy#7689]
Lint name: `needless-return`

I tried this code:
```rust
use std::net::UdpSocket;

fn get() -> Option<String> {
    let socket = match UdpSocket::bind("0.0.0.0:0") {
        Ok(s) => s,
        Err(_) => return None,
    };

    match socket.connect("8.8.8.8:80") {
        Ok(()) => (),
        Err(_) => return None,
    };

    match socket.local_addr() {
        Ok(addr) => return Some(addr.ip().to_string()),
        Err(_) => return None,
    };
}

fn main() {
    println!("{}", get().unwrap());
}
```

I expected to see this happen: no warnings/errors

Instead, this happened:

```console
error: unneeded `return` statement
  --> src/main.rs:15:21
   |
15 |         Ok(addr) => return Some(addr.ip().to_string()),
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove `return`: `Some(addr.ip().to_string())`
   |
   = note: `-D clippy::needless-return` implied by `-D warnings`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return

error: unneeded `return` statement
  --> src/main.rs:16:19
   |
16 |         Err(_) => return None,
   |                   ^^^^^^^^^^^ help: remove `return`: `None`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return
```

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.55.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.55.0
LLVM version: 12.0.1
```

**Comment by mikerite** [comment]
Why do you consider this a false positive? It looks right to me.

**Comment by yellowhat** [comment]
If I follow the advice:

```rust
use std::net::UdpSocket;

fn get() -> Option<String> {
    let socket = match UdpSocket::bind("0.0.0.0:0") {
        Ok(s) => s,
        Err(_) => return None,
    };

    match socket.connect("8.8.8.8:80") {
        Ok(()) => (),
        Err(_) => return None,
    };

    match socket.local_addr() {
        Ok(addr) => Some(addr.ip().to_string())
        Err(_) => None
    };
}

fn main() {
    println!("{}", get().unwrap());
}
```
and I get:

```console
error: expected one of `,`, `.`, `?`, `}`, or an operator, found `Err`
  --> src/main.rs:16:9
   |
15 |         Ok(addr) => Some(addr.ip().to_string())
   |                  --                            - expected one of `,`, `.`, `?`, `}`, or an operator
   |                  |
   |                  while parsing the `match` arm starting here
16 |         Err(_) => None
   |         ^^^ unexpected token
```

Am I misreading?

**Comment by mikerite** [comment]
You are misreading because the suggestion because it doesn't  say anything about deleting the commas. It's still not 100% though because you need to delete the final semi-colon as well.

The code below is what it's trying to get to.

```
use std::net::UdpSocket;

fn get() -> Option<String> {
    let socket = match UdpSocket::bind("0.0.0.0:0") {
        Ok(s) => s,
        Err(_) => return None,
    };

    match socket.connect("8.8.8.8:80") {
        Ok(()) => (),
        Err(_) => return None,
    };

    match socket.local_addr() {
        Ok(addr) => Some(addr.ip().to_string()),
        Err(_) => None,
    }
}

fn main() {
    println!("{}", get().unwrap());
}
```

This isn't a false positive but a bad suggestion.

**Comment by yellowhat** [comment]
Yes I tried that:

```rust
use std::net::UdpSocket;

fn get() -> Option<String> {
    let socket = match UdpSocket::bind("0.0.0.0:0") {
        Ok(s) => s,
        Err(_) => return None,
    };

    match socket.connect("8.8.8.8:80") {
        Ok(()) => (),
        Err(_) => return None,
    };

    match socket.local_addr() {
        Ok(addr) => Some(addr.ip().to_string()),
        Err(_) => None,
    };
}

fn main() {
    println!("{}", get().unwrap());
}
```

but I get:
```console
# cargo build --release
   Compiling aaa v0.1.0 (/data)
error[E0308]: mismatched types
  --> src/main.rs:3:13
   |
3  | fn get() -> Option<String> {
   |    ---      ^^^^^^^^^^^^^^ expected enum `Option`, found `()`
   |    |
   |    implicitly returns `()` as its body has no tail or `return` expression
...
17 |     };
   |      - help: consider removing this semicolon
   |
   = note:   expected enum `Option<String>`
           found unit type `()`
```
it will not even build.

**Comment by dswij** [comment]
There is a semicolon at the end of the last match statement.

```rust
    match socket.local_addr() {
        Ok(addr) => Some(addr.ip().to_string()),
        Err(_) => None,
    }; <--- should not be a semicolon; should be removed
```

else, it's interpreted as:

```rust
fn get() -> Option<String> {
    let socket = match UdpSocket::bind("0.0.0.0:0") {
        Ok(s) => s,
        Err(_) => return None,
    };

    match socket.connect("8.8.8.8:80") {
        Ok(()) => (),
        Err(_) => return None,
    };

    match socket.local_addr() {
        Ok(addr) => Some(addr.ip().to_string()),
        Err(_) => None,
    };

    ()
}
```

and hence, the error that you get.

Though I'd say that this is not a false positive, I agree on the point that this is bad suggestion by Clippy, something that can be improved nevertheless.

**Comment by yellowhat** [comment]
Thanks for the suggestion.
It fixed my issue.
